File: src/common/vcs/vcs.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type t =
  | Hg
  | Git

let find_root : ?recursion_limit:int -> File_path.t -> (t * File_path.t) option =
  let rec walk ~recursion_limit dir =
    if dir = File_path.parent dir then
      (* Reached fs root *)
      None
    else if File_path.file_exists (File_path.concat dir ".hg") then
      Some (Hg, dir)
    else if File_path.file_exists (File_path.concat dir ".git") then
      Some (Git, dir)
    else if recursion_limit <= 0 then
      None
    else
      walk ~recursion_limit:(recursion_limit - 1) (File_path.parent dir)
  in
  (fun ?(recursion_limit = 100) dir -> walk ~recursion_limit dir)

let find ?recursion_limit dir =
  match find_root ?recursion_limit dir with
  | Some (t, _) -> Some t
  | None -> None

let name = function
  | Git -> "Git"
  | Hg -> "Mercurial"

File: rust_port/crates/flow_common_vcs/src/vcs.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::path::Path;
use std::path::PathBuf;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Vcs {
    Hg,
    Git,
}

pub fn find_root(recursion_limit: Option<usize>, dir: &Path) -> Option<(Vcs, PathBuf)> {
    let recursion_limit = recursion_limit.unwrap_or(100);
    fn walk(recursion_limit: usize, dir: &Path) -> Option<(Vcs, PathBuf)> {
        let parent_dir = dir.parent()?;
        if dir.join(".hg").exists() {
            Some((Vcs::Hg, dir.to_path_buf()))
        } else if dir.join(".git").exists() {
            Some((Vcs::Git, dir.to_path_buf()))
        } else if recursion_limit == 0 {
            None
        } else {
            walk(recursion_limit - 1, parent_dir)
        }
    }
    walk(recursion_limit, dir)
}

pub fn find(recursion_limit: Option<usize>, dir: &Path) -> Option<Vcs> {
    find_root(recursion_limit, dir).map(|(t, _)| t)
}

pub fn name(vcs: Vcs) -> &'static str {
    match vcs {
        Vcs::Git => "Git",
        Vcs::Hg => "Mercurial",
    }
}
